File: AoC_Scala3/src/day02/day02.scala
package day02

import AoC_Lib.*

object day02 {
  val cmds: Seq[Pos] = inputStrs("input02.txt")
    .map(_.split(' '))
    .collect{ case Array(a1,a2) => a1 -> a2.toInt match {
      case ("forward",i) => Pos(i,0)
      case ("up",i) => Pos(0,-i)
      case ("down",i) => Pos(0,i)
      case _ => Pos.zero
    }}.toList

  @main
  def main(): Unit = {
    val res1 = cmds.foldLeft(Pos.zero)(_+_)
    Console.println(res1.x * res1.y) // 2215080

    val res2 = cmds.foldLeft((0,Pos.zero)){
      case ((aim, pos), cur) => cur.x match {
        case 0 => aim + cur.y -> pos
        case _ => aim -> (pos + Pos(cur.x, cur.x * aim))
      }
    }
    Console.println(res2._2.x * res2._2.y) // 1864715580
  }
}

File: Aoc_Kotlin/src/main/kotlin/day02/dive/Day02.kt
package day02.dive

import AoC_Lib.Pos
import AoC_Lib.SomeDay
import AoC_Lib.toStrs

object Day02: SomeDay(2021,2) {
    override val title = "Dive!"

    private fun String.toPos(): Pos = this
            .split(" ")
            .let { when (it[0]) {
                "forward" -> Pos(it[1].toInt(),0)
                "up" -> Pos(0, -it[1].toInt())
                "down" -> Pos(0, it[1].toInt())
                else -> Pos.Zero
            }}

    override fun first(data: String): Any? {
        return data.toStrs()
            .map { it.toPos() }
            .fold(Pos.Zero, Pos::plus)
            .let { it.x * it.y }
    } // 2215080 Time: 36ms

    override fun second(data: String): Any? {
        return data.toStrs()
            .map { it.toPos() }
            .fold(0 to Pos.Zero) {
                (aim,pos), cur ->
                    if (cur.x == 0)
                        aim + cur.y to pos
                    else
                        aim to pos + Pos(cur.x,cur.x * aim)
            }
            .let { it.second.x * it.second.y }
    } // 1864715580 Time: 3ms
}

fun main() = SomeDay.mainify(Day02)